### PR TITLE
Frontend Testing : Add test for backend Saga

### PR DESCRIPTION
**Pull request by helloImHai**
Part of #581

**Summary**:
Added test for action types FETCH_AUTH, FETCH_ASSESSMENT_OVERVIEWS, FETCH_ASSESSMENT, SUBMIT_ANSWER, SUBMIT_ASSESSMENT for backendSaga.

**Change log**:
- Changed store.dispatch to yield put() and put() in backendSaga.ts
- Refactor FETCH_AUTH, SUBMIT_ANSWER, SUBMIT_ASSESSMENT in backendSaga to make happy path prominent and use guard clauses.
- Refactor SUBMIT_ANSWER to not mutate store.

### CODE REVIEW

File: src/sagas/__tests__/backend.ts [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+          errorResp
+        ]
+      ])
+      .hasFinalState(mockStates)
+      .dispatch({ type: actionTypes.SUBMIT_ANSWER, payload: mockAnsweredAssessmentQuestion })
+      .silentRun();
+  });
+});
+
+describe('SUBMIT_ASSESSMENT ACTION DISPATCHED', () => {
```
Comment (rrtheonlyone) [inline src/sagas/__tests__/backend.ts]: Refactor the original code for this as well to use guard clauses!

File: src/sagas/backend.ts [hunk 2]
```diff
@@ -22,7 +22,7 @@ import {
   QuestionType,
   QuestionTypes
 } from '../components/assessment/assessmentShape';
-import { store } from '../createStore';
+// import { store } from '../createStore';
```
Comment (rrtheonlyone) [inline src/sagas/backend.ts]: Remove this comment?

File: src/sagas/__tests__/backend.ts [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+};
+
+const okResp = { ok: true };
+const errorResp = { ok: false };
+// ----------------------------------------
+
+describe('FETCH_AUTH ACTION DISPATCHED', () => {
+  test('and backendSaga runs as intended', () => {
+    const luminousCode = 'luminousCode';
+    const user = mockTokens ? 'user' : null;
```
[the remark this answers is not in the document]
Comment (helloImHai) [reply src/sagas/__tests__/backend.ts]: It's to get around a type error I had, but then doing refactoring like your comment below gets rid of it.

File: src/sagas/__tests__/backend.ts [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+      .hasFinalState(mockStates)
+      .dispatch({ type: actionTypes.FETCH_ASSESSMENT, payload: mockId })
+      .silentRun();
+  });
+});
+
+describe('SUBMIT_ANSWER ACTION DISPATCHED', () => {
+  test('and saga runs as intended', () => {
+    const mockAnsweredAssessmentQuestion = { ...mockAssessmentQuestion, answer: '42' };
+    const mockNewQuestions = mockAssessment.questions.slice().map((question: IQuestion) => {
```
[the remark this answers is not in the document]
Comment (helloImHai) [reply src/sagas/__tests__/backend.ts]: I don't really understand this comment, you mean that mockAssessmentQuestion which is inside the assessment state will change?

[the remark this answers is not in the document]
Comment (helloImHai) [reply src/sagas/__tests__/backend.ts]: Yep the prior way of question.answer = answer of the backend.ts did mutate state, I think this fixes it. I put in the expect at the end to make sure.

File: src/sagas/__tests__/backend.ts [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+    const luminousCode = 'luminousCode';
+    const user = 'user';
+    return expectSaga(backendSaga)
+      .put(actions.setTokens(mockTokens))
+      .put(actions.setUser(user))
+      .provide([[call(postAuth, luminousCode), mockTokens], [call(getUser, mockTokens), user]])
+      .dispatch({ type: actionTypes.FETCH_AUTH, payload: luminousCode })
+      .silentRun();
+  });
+  test('when tokens is null', () => {
```
Comment (rrtheonlyone) [inline src/sagas/__tests__/backend.ts]: This test is actually doing nothing. You should assert that `setTokens` / `setUser` is not being called.

Refer to https://redux-saga-test-plan.jeremyfairbank.com/integration-testing/negated-assertions.html

File: src/sagas/__tests__/backend.ts [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+    const luminousCode = 'luminousCode';
+    const user = 'user';
+    return expectSaga(backendSaga)
+      .provide([[call(postAuth, luminousCode), null], [call(getUser, mockTokens), user]])
+      .dispatch({ type: actionTypes.FETCH_AUTH, payload: luminousCode })
+      .silentRun();
+  });
+  test('when user is null', () => {
```
Comment (rrtheonlyone) [inline src/sagas/__tests__/backend.ts]: same for this

File: src/sagas/__tests__/backend.ts [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+    return expectSaga(backendSaga)
+      .withState({ session: mockTokens })
+      .provide([[call(getAssessmentOverviews, mockTokens), mockAssessmentOverviews]])
+      .put(actions.updateAssessmentOverviews(mockAssessmentOverviews))
+      .hasFinalState({ session: mockTokens })
+      .dispatch({ type: actionTypes.FETCH_ASSESSMENT_OVERVIEWS })
+      .silentRun();
+  });
+
+  test('when assesments is null', () => {
```
Comment (rrtheonlyone) [inline src/sagas/__tests__/backend.ts]: same for this

File: src/sagas/__tests__/backend.ts [hunk 8]
```diff
@@ -0,0 +1,265 @@
...
+    return expectSaga(backendSaga)
+      .withState({ session: mockTokens })
+      .provide([[call(getAssessment, mockId, mockTokens), mockAssessment]])
+      .put(actions.updateAssessment(mockAssessment))
+      .hasFinalState({ session: mockTokens })
+      .dispatch({ type: actionTypes.FETCH_ASSESSMENT, payload: mockId })
+      .silentRun();
+  });
+
+  test('when assesment is null', () => {
```
Comment (rrtheonlyone) [inline src/sagas/__tests__/backend.ts]: and this

File: src/sagas/__tests__/backend.ts [hunk 9]
```diff
@@ -0,0 +1,265 @@
...
+    const mockAnsweredAssessmentQuestion = { ...mockAssessmentQuestion, answer: '42' };
+    return expectSaga(backendSaga)
+      .withState({ session: { role: Role.Staff } })
+      .call(showWarningMessage, 'Only students can submit answers.')
+      .hasFinalState({ session: { role: Role.Staff } })
+      .dispatch({ type: actionTypes.SUBMIT_ANSWER, payload: mockAnsweredAssessmentQuestion })
+      .silentRun();
+  });
+
+  test('when respond is null', () => {
```
Comment (rrtheonlyone) [inline src/sagas/__tests__/backend.ts]: when response is null

File: src/sagas/__tests__/backend.ts [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+      ])
+      .put(actions.updateAssessment(mockNewAssessment))
+      .put(actions.updateHasUnsavedChanges('assessment' as WorkspaceLocation, false))
+      .dispatch({ type: actionTypes.SUBMIT_ANSWER, payload: mockAnsweredAssessmentQuestion })
+      .silentRun();
+    // To make sure no changes in state
+    return expect(mockStates.session.assessments.get(0)!.questions[0].answer).toEqual(null);
+  });
+
+  test('when role is not student', () => {
```
Comment (rrtheonlyone) [inline src/sagas/__tests__/backend.ts]: check that update is not called for all the null/error cases as well!

File: src/sagas/backend.ts [hunk 11]
```diff
@@ -608,7 +612,8 @@ async function request(
       return response;
     } else if (opts.shouldRefresh && response.status === 401) {
       const newTokens = await postRefresh(opts.refreshToken!);
-      store.dispatch(actions.setTokens(newTokens));
+      // Changed put to put
```
Comment (rrtheonlyone) [inline src/sagas/backend.ts]: What does comment mean?

File: src/sagas/backend.ts [hunk 12]
```diff
@@ -624,7 +629,8 @@ async function request(
       throw new Error('API call failed or got non-OK response');
     }
   } catch (e) {
-    store.dispatch(actions.logOut());
+    // Changed put to put
```
Comment (rrtheonlyone) [inline src/sagas/backend.ts]: Same here!

**Pull request merged**